src/bio/bio_buffer.c

```diff
@@ -446,14 +446,21 @@ struct bio_copy_args {
 };
 
 static int
-copy_one(struct bio_desc *biod, struct bio_iov *biov, void *data)
+copy_one(struct bio_desc *biod, struct bio_iov *biov, int idx, void *data)
 {
 	struct bio_copy_args	*arg = data;
 	d_sg_list_t		*sgl;
 	void			*addr = bio_iov2req_buf(biov);
 	ssize_t			 size = bio_iov2req_len(biov);
 	uint16_t		 media = bio_iov2media(biov);
 
+	D_ASSERT(idx < arg->ca_sgl_cnt);
+	arg->ca_sgl_idx = idx;
+	arg->ca_iov_idx = 0;
+	arg->ca_iov_off = 0;
+	if (biod->bd_type == BIO_IOD_TYPE_FETCH)
+		arg->ca_sgls[idx].sg_nr_out = 0;
```

Comment:
This per SGL stuff should only be initialized once (when switching to a new SGL), but the callback is called for each IOV, so it's not correct to do such initialization work in every callback.

So maybe the callback could be modified to accept both "sgl_idx" and "iov_idx" as parameter, and only perform the per SGL initialization work when "iov_idx == 0"?

Reply:
I'm not sure even that would work as what about if bs_nr_out is zero?  The callback isn't invoked at all in that case?

I wonder about having two callbacks, one for sgls and one for iovs?